src/osd/PrimaryLogPG.cc

```diff
@@ -2489,6 +2489,11 @@ PrimaryLogPG::cache_result_t PrimaryLogPG::maybe_handle_cache_detail(
     osd->logger->inc(l_osd_op_cache_hit);
     return cache_result_t::NOOP;
   }
+  if (!is_primary()) {
+    dout(20) << __func__ << " cache miss; ask the primary" << dendl;
+    osd->reply_op_error(op, -EAGAIN);
+    return cache_result_t::REPLIED_WITH_EAGAIN;
```

Comment:
I think we actually need this to go at the top or we might get weird ordering bugs. eg, if the primary is promoting objects while the listing is happening, or if we return out for NOOPs because we have incorrect metadata that we're looking at as a result of being non-primary.

Reply:
The replica will have a missing set, so it'll bail out already if the pool is out of sync for the object in question before this method is called in do_op().  And more practically we can't move it up any higher in this method anyway or else we're above the "cache hit" case.  Do you have a specific case in mind?  I can't think of anything..

Reply:
Well, I don't remember anything more specific now. I guess if we're a replica we're only serving read operations, so as long as we can't serve a read while the object is somehow invalid then we're okay.
I can't think of any way the primary could logically finish nuking or overwriting an object that we wouldn't know about as a replica, so we should be good. Can you?